File: packages/reference-engine/src/conditions.js
'use strict';
/**
 * 过滤条件算子的参考实现,对应 docs/reference/operators.md §3。
 *
 * 规格要求:**下表全部算子均有实现**。1.x 声明了完整算子集但引擎只实现了
 * 其中一小部分,用户配得出来的变量跑不了 —— 这正是 2.0 引入强制校验的动因。
 *
 * 两条贯穿性约定:
 *   - 类型严格:实际类型与声明类型不符时判定为**不通过**,不做隐式转换
 *   - null 输入判定为不通过(empty/!empty 除外,它们就是用来判空的)
 */

const REGEX_CACHE = new Map();
function compileRegex(pattern) {
  if (!REGEX_CACHE.has(pattern)) {
    // 规格:整串匹配语义
    REGEX_CACHE.set(pattern, new RegExp(`^(?:${pattern})$`));
  }
  return REGEX_CACHE.get(pattern);
}

/** 部分匹配语义的正则(1.x 的 !regex 用于判空,如 `^\s*$`,需要能匹配整串) */
function testRegex(pattern, value) {
  return compileRegex(pattern).test(value);
}

function asString(v) {
  return typeof v === 'string' ? v : null;
}
function asNumber(v) {
  return typeof v === 'number' && Number.isFinite(v) ? v : null;
}

/** 地理位置查询由调用方注入,缺省实现返回 unknown(规格 §3.4) */
let locationResolver = () => 'unknown';
function setLocationResolver(fn) { locationResolver = fn; }

const CONDITIONS = {
  // ---------- 通用 ----------
  '==': (a, b) => looseEq(a, b),
  '!=': (a, b) => !looseEq(a, b),

  empty: (a) => a === null || a === undefined || a === '',
  '!empty': (a) => !(a === null || a === undefined || a === ''),

  // ---------- 数值 ----------
  '>': (a, b) => cmpNum(a, b, (x, y) => x > y),
  '>=': (a, b) => cmpNum(a, b, (x, y) => x >= y),
  '<': (a, b) => cmpNum(a, b, (x, y) => x < y),
  '<=': (a, b) => cmpNum(a, b, (x, y) => x <= y),

  // ---------- 字符串 ----------
  contains: (a, b) => strOp(a, b, (x, y) => x.includes(y)),
  '!contains': (a, b) => negate(strOp(a, b, (x, y) => x.includes(y))),
  startwith: (a, b) => strOp(a, b, (x, y) => x.startsWith(y)),
  '!startwith': (a, b) => negate(strOp(a, b, (x, y) => x.startsWith(y))),
  endwith: (a, b) => strOp(a, b, (x, y) => x.endsWith(y)),
  '!endwith': (a, b) => negate(strOp(a, b, (x, y) => x.endsWith(y))),
  regex: (a, b) => strOp(a, b, (x, y) => testRegex(y, x)),
  '!regex': (a, b) => negate(strOp(a, b, (x, y) => testRegex(y, x))),
  // containsby:字段值是给定值的子串(反向包含)
  containsby: (a, b) => strOp(a, b, (x, y) => y.includes(x)),
  '!containsby': (a, b) => negate(strOp(a, b, (x, y) => y.includes(x))),
  in: (a, b) => inSet(a, b),
  '!in': (a, b) => negate(inSet(a, b)),

  // ---------- IP 地理 ----------
  locationequals: (a, b, ctx) => {
    const loc = locationResolver(a, (ctx && ctx.level) || 'province');
    return loc === b;
  },
  '!locationequals': (a, b, ctx) => !CONDITIONS.locationequals(a, b, ctx),
  locationcontainsby: (a, b, ctx) => {
    const loc = locationResolver(a, (ctx && ctx.level) || 'province');
    return splitSet(b).includes(loc);
  },
  '!locationcontainsby': (a, b, ctx) => !CONDITIONS.locationcontainsby(a, b, ctx),
};

function looseEq(a, b) {
  if (a === null || a === undefined) return false;
  // 1.x 的常量一律以字符串存储,比较时按左值类型对齐(规格「常量值保持字符串」)
  if (typeof a === 'number') {
    const n = typeof b === 'number' ? b : Number(b);
    return Number.isFinite(n) && a === n;
  }
  if (typeof a === 'boolean') return a === (b === true || b === 'true');
  return String(a) === String(b);
}

function cmpNum(a, b, fn) {
  const x = asNumber(a);
  if (x === null) return false;                 // 类型严格
  const y = typeof b === 'number' ? b : Number(b);
  if (!Number.isFinite(y)) return false;
  return fn(x, y);
}

function strOp(a, b, fn) {
  const x = asString(a);
  if (x === null) return false;                 // 类型严格
  return fn(x, String(b));
}

function negate(v) { return !v; }

function splitSet(b) {
  if (Array.isArray(b)) return b.map(String);
  return String(b).split(',').map((s) => s.trim());
}

function inSet(a, b) {
  if (a === null || a === undefined) return false;
  return splitSet(b).includes(String(a));
}

/**
 * 求值一个过滤条件树(variable-model 的 filter 结构)。
 * 复合条件短路求值:and 遇 false 即返回,or 遇 true 即返回,not 只取第一个。
 */
function evalFilter(filter, getField) {
  if (!filter || Object.keys(filter).length === 0) return true;   // 空对象表示不过滤

  const type = filter.type || 'simple';
  if (type === 'simple') {
    const op = CONDITIONS[filter.operation];
    if (!op) {
      throw new Error(`未实现的条件算子: ${filter.operation}(规格见 docs/reference/operators.md §3)`);
    }
    return !!op(getField(filter.object), filter.value, filter);
  }
  const subs = filter.condition || [];
  if (type === 'and') {
    for (const s of subs) if (!evalFilter(s, getField)) return false;
    return true;
  }
  if (type === 'or') {
    for (const s of subs) if (evalFilter(s, getField)) return true;
    return false;
  }
  if (type === 'not') {
    return !evalFilter(subs[0], getField);
  }
  throw new Error(`未知的条件组合类型: ${type}`);
}

module.exports = { CONDITIONS, evalFilter, setLocationResolver };

File: apps/engine/src/main/java/cn/threathunter/nebula/engine/condition/Conditions.java
package cn.threathunter.nebula.engine.condition;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

/**
 * 过滤与比较条件算子,对应 {@code docs/reference/operators.md} §3。
 *
 * <p>规格要求<b>下表全部算子均有实现</b>。1.x 声明了完整算子集但引擎只实现了其中
 * 一小部分(字符串仅 contains/==/!=,Double 仅 &lt;),用户配得出来的变量跑不了 ——
 * 这正是 2.0 引入 schema 强制校验的直接动因。
 *
 * <p>两条贯穿性约定:
 * <ul>
 *   <li><b>类型严格</b>:实际类型与声明类型不符时判定为不通过,不做隐式转换</li>
 *   <li><b>null 判定为不通过</b>(empty / !empty 除外,它们就是用来判空的)</li>
 * </ul>
 */
public final class Conditions {

    private Conditions() {
    }

    /** IP 归属地查询。缺省返回 unknown,与规格 §3.4 一致(查询失败不使条件报错)。 */
    @FunctionalInterface
    public interface LocationResolver {
        String resolve(String ip, String level);
    }

    private static volatile LocationResolver locationResolver = (ip, level) -> "unknown";

    public static void setLocationResolver(LocationResolver r) {
        locationResolver = r == null ? (ip, level) -> "unknown" : r;
    }

    private static final Map<String, Pattern> REGEX_CACHE = new LinkedHashMap<>();

    private static Pattern compile(String pattern) {
        synchronized (REGEX_CACHE) {
            return REGEX_CACHE.computeIfAbsent(pattern, p -> {
                try {
                    return Pattern.compile("^(?:" + p + ")$"); // 规格:整串匹配语义
                } catch (PatternSyntaxException e) {
                    throw new IllegalArgumentException("条件正则无法编译: " + p, e);
                }
            });
        }
    }

    /** 求值单个条件。 */
    public static boolean eval(String op, Object left, Object right) {
        if (op == null) {
            throw new IllegalArgumentException("条件算子为空");
        }
        return switch (op) {
            case "==" -> looseEquals(left, right);
            case "!=" -> !looseEquals(left, right);

            case "empty" -> left == null || "".equals(left);
            case "!empty" -> !(left == null || "".equals(left));

            case ">" -> compareNum(left, right, c -> c > 0);
            case ">=" -> compareNum(left, right, c -> c >= 0);
            case "<" -> compareNum(left, right, c -> c < 0);
            case "<=" -> compareNum(left, right, c -> c <= 0);

            case "contains" -> strOp(left, right, (a, b) -> a.contains(b));
            case "!contains" -> !strOp(left, right, (a, b) -> a.contains(b));
            case "startwith" -> strOp(left, right, String::startsWith);
            case "!startwith" -> !strOp(left, right, String::startsWith);
            case "endwith" -> strOp(left, right, String::endsWith);
            case "!endwith" -> !strOp(left, right, String::endsWith);
            case "regex" -> strOp(left, right, (a, b) -> compile(b).matcher(a).matches());
            case "!regex" -> !strOp(left, right, (a, b) -> compile(b).matcher(a).matches());
            // containsby:反向包含 —— 字段值是给定值的子串
            case "containsby" -> strOp(left, right, (a, b) -> b.contains(a));
            case "!containsby" -> !strOp(left, right, (a, b) -> b.contains(a));
            case "in" -> inSet(left, right);
            case "!in" -> !inSet(left, right);

            case "locationequals" -> locationOf(left).equals(String.valueOf(right));
            case "!locationequals" -> !locationOf(left).equals(String.valueOf(right));
            case "locationcontainsby" -> splitSet(right).contains(locationOf(left));
            case "!locationcontainsby" -> !splitSet(right).contains(locationOf(left));

            default -> throw new IllegalArgumentException(
                    "未实现的条件算子: " + op + "(规格见 docs/reference/operators.md §3)");
        };
    }

    private static String locationOf(Object ip) {
        if (ip == null) {
            return "unknown";
        }
        String r = locationResolver.resolve(String.valueOf(ip), "province");
        return r == null || r.isEmpty() ? "unknown" : r;
    }

    /**
     * 相等比较。1.x 的常量一律以字符串存储(阈值 5 存为 "5"),因此比较时按
     * <b>左值类型</b>对齐 —— 这与迁移工具「常量值保持字符串、由引擎按左值类型
     * 转换」的约定配套。
     */
    private static boolean looseEquals(Object left, Object right) {
        if (left == null) {
            return false;
        }
        if (left instanceof Number ln) {
            Double rn = asNumber(right);
            return rn != null && ln.doubleValue() == rn;
        }
        if (left instanceof Boolean lb) {
            return lb == (Boolean.TRUE.equals(right) || "true".equals(String.valueOf(right)));
        }
        return String.valueOf(left).equals(String.valueOf(right));
    }

    private static Double asNumber(Object o) {
        if (o instanceof Number n) {
            return n.doubleValue();
        }
        if (o instanceof String s) {
            try {
                return Double.parseDouble(s.trim());
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return null;
    }

    private interface IntPred {
        boolean test(int cmp);
    }

    private static boolean compareNum(Object left, Object right, IntPred pred) {
        if (!(left instanceof Number ln)) {
            return false; // 类型严格:非数值一律不通过,不做隐式转换
        }
        Double rn = asNumber(right);
        if (rn == null) {
            return false;
        }
        return pred.test(Double.compare(ln.doubleValue(), rn));
    }

    private interface StrPred {
        boolean test(String a, String b);
    }

    private static boolean strOp(Object left, Object right, StrPred pred) {
        if (!(left instanceof String s)) {
            return false; // 类型严格
        }
        return pred.test(s, String.valueOf(right));
    }

    private static List<String> splitSet(Object right) {
        if (right instanceof List<?> list) {
            List<String> out = new ArrayList<>();
            list.forEach(x -> out.add(String.valueOf(x)));
            return out;
        }
        String s = String.valueOf(right);
        List<String> out = new ArrayList<>();
        for (String p : s.split(",")) {
            out.add(p.trim());
        }
        return out;
    }

    private static boolean inSet(Object left, Object right) {
        if (left == null) {
            return false;
        }
        return splitSet(right).contains(String.valueOf(left));
    }

    /** 全部已实现的算子,供 schema 一致性检查使用。 */
    public static List<String> supported() {
        return Arrays.asList(
                "==", "!=", "empty", "!empty",
                ">", ">=", "<", "<=",
                "contains", "!contains", "startwith", "!startwith",
                "endwith", "!endwith", "regex", "!regex",
                "containsby", "!containsby", "in", "!in",
                "locationequals", "!locationequals",
                "locationcontainsby", "!locationcontainsby");
    }
}
